### PR TITLE
Continuing work on expressions

### PR DESCRIPTION
**Pull request by aalkin**
* Reworked tree-parsing algorithm to use ArrowDatumSpec
* Added a test of tree-parsing algorithm

### CODE REVIEW

File: Framework/Core/include/Framework/Expressions.h [hunk 1]
```diff
@@ -11,10 +11,12 @@
 #define O2_FRAMEWORK_EXPRESSIONS_H_
 
 #include "Framework/Kernels.h"
+#include "Framework/VariantHelpers.h"
 
 #include <variant>
 #include <string>
 #include <memory>
+#include <iostream>
```
Comment (ktf) [inline Framework/Core/include/Framework/Expressions.h]: Never include <iostream> in an header, use <iosfwd>.

**Comment by aalkin** [comment]
There is QualityControl build log (with error) included in the failing build logs, is it supposed to be there?

**Comment by ktf** [comment]
No, but it's there everywhere since yesterday. I could not yet understand what changed yet.

**Comment by aalkin** [comment]
I see that the template instantiation errors in the QC log are definitely connected with Expressions: it picks up the variant from inside ArrowDatumSpec, which is `variant<monostate,size_t,variant<int,bool.float,double>,string>`, but I can not determine in which context.

**Comment by ktf** [comment]
The problem itself is due to the fact that the clang used by cling (5.0.2) has a known bug when using GCC headers < 8. The way to fix it is to hide whatever uses variant from cling. Why do you need to expose the ArrowDatum in Expression? Can you simply put an `#ifdef __CLING__` around the variant and the header?

File: Framework/Core/include/Framework/Expressions.h [hunk 2]
```diff
@@ -11,10 +11,12 @@
 #define O2_FRAMEWORK_EXPRESSIONS_H_
 
 #include "Framework/Kernels.h"
+#include "Framework/VariantHelpers.h"
```
Comment (ktf) [inline Framework/Core/include/Framework/Expressions.h]: This should probably put only in the .cxx.

**Comment by aalkin** [comment]
@ktf I moved the parsing helpers into a separate header that is not exposed in include, that should hide variant from cling.

**Comment by ktf** [comment]
Errors unrelated. This hides variant from ROOT.

**Pull request merged**